rfc/rfc-42/rfc-42.md

```diff
@@ -0,0 +1,230 @@
+Though it is not required in a standard Consistent Hashing algorithm, fragmentations in hash ranges may cause extra complexity for the splitting process in our case.
+
+**For splitting files**, a split point (i.e., hash ranges for the output buckets) should be decided:
+
+- A simple policy would be split in the range middle, but it may produce uneven data files.
+In an extreme case, splitting may produce one file with all data and one file with no data.
+- Another policy is to find a split point that evenly dispatches records into children buckets.
+It requires knowledge about the hash value distribution of the original buckets.
+
+In our implementation, *we will first stick to the first simple one*, as buckets will finally converge to a balanced distribution after multiple rounds of resizing processes.
+Of course, a pluggable implementation will be kept for extensibility so that users can choose different available policies.
+
+All updates related to the hash metadata will be first recorded in the clustering plan, and then be reflected in partitions' hashing metadata when clustering finishes.
+The plan is generated and persisted in files during the scheduling process, which is protected by a table-level lock for a consistent table view.
+
+### Concurrent Writer & Reader
+
+Concurrent updates to file groups that are under clustering are not supported in Hudi, since records location change during the re-organizing.
+For example, a clustering process is scheduled to combine three parquet files A, B, C into two parquet files E, F with a more compact storage layout.
+During the clustering, a writer comes in and plan to update a record that originally belongs to the parquet file A.
+There are two choices the writer can make:
+
+- Only update the old file (i.e. parquet file A), but the update will lost once the clustering finishes because of the *replacecommit*'s shadowing.
+- Besides parquet file A, also update new files (i.e. parquet file E or F).
+However, the record's new location can only be determined until the clustering finishes, which means the writer will be blocked as it has to wait for the location information.
+
+For tables using Bucket Index, the above conflicting condition can be avoided because record locations are calculated by the hash algorithm.
+So even before the clustering finishes, the writer can calculate record locations as long as it knows the latest hash algorithm (in our Consistent Hashing case, it is stored in the clustering plan).
+
+#### Dual write solution
+
+![bucket resizing](bucket_resizing.png)
+
+The figure above shows a concurrent write process during the bucket resizing process (i.e., splitting).
+Dual-write is activated for buckets that are splitting.
+Each incoming record will be written to log files of both the old file group (i.e., FG1) and children file groups (i.e., FG1.1 or FG1.2).
```

Comment:
so, with this, a regular writer could end up initializing the new file group is it? as of today (w/o this feature), new file groups for clustering is never touched by regular writers and so, initialization does not need to happen upfront. But with the proposal, if clustering/resizing is scheduled, but new files are not initialized, and if a regular writes comes through, will it take care of instantiating the file group?
If yes, do we guard the initialization using a lock? if not, what happens if there are two concurrent writes trying to write to a new file group that is not initialized yet?

Reply:
By 'initializing the new file group', do you mean generating file id? During the scheduling of the clustering, we have generated file group id and the new hashing mapping. So when the concurrent writer arrives, it can see all the necessary infos to operate the dual write. By the way, in the implementation, we only allow the clustering to be scheduled when there is no inflight writers, so that all writers can see a consistent table states (with or without resizing/clustering task in the background).

Also, concurrent cases also happen when multiple writers try to write a fresh new table (or a table has not been created yet). In this case, we rely on that only one client can write the hashing metadata successfully (the hashing metadata will have an initial name 00000000.hashing_meta). Other writers that failed to create the hashing metadata will wait util they can see the hashing metadata (because there maybe visibility latency of the underlying object store). Concurrency control will be carried out as usual, which means only one can commit when writings happens to the same file group.